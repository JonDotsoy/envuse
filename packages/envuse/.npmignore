**/*.ts
!**/*.d.ts
!**/*.js
__demo__
__snapshots__
*.tgz
.editorconfig
tsconfig.json
jest.config.js
**/*.spec.ts
**/*.spec.d.ts
**/*.spec.js
**/*.spec.js.map
.environment/
.__environment_demo__/
.vscode/
scripts/
