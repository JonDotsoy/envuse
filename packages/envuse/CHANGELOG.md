# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [3.3.0-beta.2](https://github.com/JonDotsoy/envuse/compare/envuse@v3.3.0-beta.1...envuse@v3.3.0-beta.2) (2021-12-12)

### Features

- **envuse:** configure cli ([67802ae](https://github.com/JonDotsoy/envuse/commit/67802ae60db3d9526016e15a12ecb8a09bf16aed))

## [3.3.0-beta.1](https://github.com/JonDotsoy/envuse/compare/envuse@v3.3.0-beta.0...envuse@v3.3.0-beta.1) (2021-12-12)

### Bug Fixes

- **envuse:** requires paths ([24e0448](https://github.com/JonDotsoy/envuse/commit/24e0448a2d6b4ea3ffa577b62081aa4582a1c96b))

## [3.3.0-beta.0](https://github.com/JonDotsoy/envuse/compare/envuse@v3.2.0...envuse@v3.3.0-beta.0) (2021-12-12)

### Features

- **envuse:** initialize envuse cli ([d79a603](https://github.com/JonDotsoy/envuse/commit/d79a603164e9725a96df5ff20da7a5a21dcae461))
- **envuse:** preapre publish beta ([f67d3eb](https://github.com/JonDotsoy/envuse/commit/f67d3eb6fa13f93ee9ac3eedfd229e41521c25a4))
- **vscode-language:** remove icon themes ([332720a](https://github.com/JonDotsoy/envuse/commit/332720adecaae00e6f8cbbc3d570d7891a534789))
- **vscode-language:** upgrade description package ([2d60c02](https://github.com/JonDotsoy/envuse/commit/2d60c02b446088f699a1c4b2da2a42380908d039))

## [3.2.0](https://github.com/JonDotsoy/envuse/compare/envuse@v3.1.0...envuse@v3.2.0) (2021-12-12)

### Features

- **envuse:** update test package envuse ([d0f3da7](https://github.com/JonDotsoy/envuse/commit/d0f3da7ff78cb0706a9e6fe8f010f14d3a5b9aa7))

## [3.1.0](https://github.com/JonDotsoy/envuse/compare/envuse@v3.0.2...envuse@v3.1.0) (2021-12-12)

### Features

- configure error class ([e2e6952](https://github.com/JonDotsoy/envuse/commit/e2e6952312cc1f00da4a9285e3a3ae5420795599))
- configure function delete file ([753461e](https://github.com/JonDotsoy/envuse/commit/753461e7c2bb1617dddfd27a9751235cb4b663d5))
- export provider ([6a5cfb6](https://github.com/JonDotsoy/envuse/commit/6a5cfb630cac0233475a1f6c15188863c1968246))
- **file-service:** declare require resource id ([baf084c](https://github.com/JonDotsoy/envuse/commit/baf084c8fb896ee93a4c916ab46a8da56d8148c9))
- public @envuse/files-service ([d5e8c6f](https://github.com/JonDotsoy/envuse/commit/d5e8c6ff3b06f85fdc2904e2409eb0fc6d658d30))

### Bug Fixes

- add dep @types/jest ([7dd4fb6](https://github.com/JonDotsoy/envuse/commit/7dd4fb6bc255fc1a28a715cce7990e225c4cebbe))
- add options configurable ([94b98cd](https://github.com/JonDotsoy/envuse/commit/94b98cd5dc4bd5f06408d442bc0bd0928bbcf757))
- add spec files to npm ignore ([b2264f7](https://github.com/JonDotsoy/envuse/commit/b2264f7b900d2282cb9780fcc25eb1905ac206a1))
- update validation type ObjectResource ([7912541](https://github.com/JonDotsoy/envuse/commit/791254197cc66e12d07d091d7759fbb612f6819d))

### [3.0.2](https://github.com/JonDotsoy/envuse/compare/envuse@v3.0.0...envuse@v3.0.2) (2021-10-16)

### Bug Fixes

- **envuse:** correction imports ([dde3c9f](https://github.com/JonDotsoy/envuse/commit/dde3c9f3060d7c3440a507032883c361c2ce2521))
- **envuse:** fix name module ctx ([8d83a1a](https://github.com/JonDotsoy/envuse/commit/8d83a1a06730cb85c2e1e08fe22d908b78e5f26e))

### [3.0.1](https://github.com/JonDotsoy/envuse/compare/envuse@v3.0.0...envuse@v3.0.1) (2021-10-16)

### Bug Fixes

- **envuse:** correction imports ([dde3c9f](https://github.com/JonDotsoy/envuse/commit/dde3c9f3060d7c3440a507032883c361c2ce2521))

## [3.0.0](https://github.com/JonDotsoy/envuse/compare/envuse@v2.4.5...envuse@v3.0.0) (2021-10-16)

### âš  BREAKING CHANGES

- **envuse:** refactor module envuse

### Features

- **envuse:** active dialog deprecation ([35045dc](https://github.com/JonDotsoy/envuse/commit/35045dca010e609c1c71c822553cecd1105e8fc4))
- **envuse:** add module load ([287b3da](https://github.com/JonDotsoy/envuse/commit/287b3daf982c782f3e217dec1f7b12f3747ddc87))
- **envuse:** add support to envuse selector ([41a11c3](https://github.com/JonDotsoy/envuse/commit/41a11c3dd41d71c4d284dc2ecfd36d533c5a77e9))
- **envuse:** configure cache available in register ([5c7a3bf](https://github.com/JonDotsoy/envuse/commit/5c7a3bf0af7cd1591e07f01eb78e03de23de702a))
- **envuse:** configure require with global config definitions ([d2fe64d](https://github.com/JonDotsoy/envuse/commit/d2fe64d19d3aa749ff086bfebc57f50b354e4ddf))
- **envuse:** configure type definitions ([845ed34](https://github.com/JonDotsoy/envuse/commit/845ed342cc663f996717562cf5ad078b6a7cf471))
- **envuse:** declare function deprecated ([a6ede0b](https://github.com/JonDotsoy/envuse/commit/a6ede0bed8934264de7d98674e70b57921fd2f81))
- **envuse:** depreacate read variable parsed ([c66355e](https://github.com/JonDotsoy/envuse/commit/c66355e643316fd21af54bbab05a0e29b5aecdef))
- **envuse:** implement cache system ([71ed2f4](https://github.com/JonDotsoy/envuse/commit/71ed2f478faa0dea3ccb0e6cb604ce6bc5092f4a))
- **envuse:** implement cache to load data ([516d53c](https://github.com/JonDotsoy/envuse/commit/516d53c3360d0e3886f440dde610398ca1fb8633))
- **envuse:** implement load data and load data sync ([5478065](https://github.com/JonDotsoy/envuse/commit/547806542eab000c448ebe332f90e7a22e0b3eeb))
- **envuse:** initialize module configure ([7725f51](https://github.com/JonDotsoy/envuse/commit/7725f51b612397707f8a22ff6d2ad1ee0966bb12))
- **envuse:** update function load and load-sync ([8f5bb1c](https://github.com/JonDotsoy/envuse/commit/8f5bb1ce0efffe7aca7f3f572f9b4872e9457cbe))
- **envuse:** update register method ([2f55cb2](https://github.com/JonDotsoy/envuse/commit/2f55cb2f3f490cd7793125b08a5e808de640932c))
- **envuse:** update types ([0db1f86](https://github.com/JonDotsoy/envuse/commit/0db1f868347353c816f59ecc68a8f8c3a4c24cf6))
- update brand ([87a6125](https://github.com/JonDotsoy/envuse/commit/87a612561f609afe78649d513abe4038aebb4d61))

### Bug Fixes

- **envuse:** update deps by audit ([6184821](https://github.com/JonDotsoy/envuse/commit/6184821e2f96f01856a5e27d6a97b3a3adffb6a4))

- **envuse:** refactor module envuse ([cff8c61](https://github.com/JonDotsoy/envuse/commit/cff8c61e14f2730a0b8508f4b5cefb9fbab7e29b))

### [2.4.5](https://github.com/JonDotsoy/envuse/compare/envuse@v2.4.4...envuse@v2.4.5) (2021-07-29)

### Features

- **envuse:** upgrade README file ([684897b](https://github.com/JonDotsoy/envuse/commit/684897b4be7c559365cd135e6f1995fa5e27a054))

### [2.4.4](https://github.com/JonDotsoy/envuse/compare/envuse@v2.4.3...envuse@v2.4.4) (2021-07-29)

### Features

- **envuse:** upgrade readme file ([09ffb82](https://github.com/JonDotsoy/envuse/commit/09ffb823e7ead332f068630c80c483fc88afb2a5))
- **package:** rename package envuse-vscode-syntaxes - ([1c7e1bc](https://github.com/JonDotsoy/envuse/commit/1c7e1bc122277d3e45172cfeb9257e74c9d68bf8))
- **package:** upgrade name package vsce language ([cfa7ddd](https://github.com/JonDotsoy/envuse/commit/cfa7ddd931151b90dd84e9fd8e502ce2dcb84ee4))
- **vscode-language:** configure extension icon ([3b1688d](https://github.com/JonDotsoy/envuse/commit/3b1688db84ea8764705e7b8a37fdfb6a425e3de3))
- **vscode-language:** configure post release with publish vsce ([e857bae](https://github.com/JonDotsoy/envuse/commit/e857bae0cfacb5789ff499ca4bc034614f9ae11e))
- **vscode-language:** upgrade script release ([64ee828](https://github.com/JonDotsoy/envuse/commit/64ee8286e1665190b989de7161c3099e27501d1c))

### [2.4.3](https://github.com/JonDotsoy/envuse/compare/envuse@v2.4.2...envuse@v2.4.3) (2021-07-29)

### Features

- **envuse:** upgrade version as latest ([aa98f1f](https://github.com/JonDotsoy/envuse/commit/aa98f1f835687a8dcdcaf6bfa69bee7e657c380f))
- **package:** add script release envuse ([c2f01fa](https://github.com/JonDotsoy/envuse/commit/c2f01fa4eed656e4506e7295483aa6195e97184f))

### [2.4.2](https://github.com/JonDotsoy/envuse/compare/envuse@v2.4.1...envuse@v2.4.2) (2021-07-29)

### Features

- **envuse:** add folder environment demo on gitignore file ([675b7af](https://github.com/JonDotsoy/envuse/commit/675b7afe836c7c9087b157a7fa2d8ae8323e3b35))
- **envuse:** add imagen demo to envuse file ([65991b4](https://github.com/JonDotsoy/envuse/commit/65991b41aa261db0e672479f1910adcccc72c6c0))
- **envuse:** update ([c664fdf](https://github.com/JonDotsoy/envuse/commit/c664fdf1483e9b6511fca46bd601fd82fc5e1195))
- **envuse:** upgrade sample comment ([6ed67ba](https://github.com/JonDotsoy/envuse/commit/6ed67ba9ce5678f0489b84c73e515d344adab4d8))
- **envuse:** upgrade tests and rename comment to comment inline ([3ea955b](https://github.com/JonDotsoy/envuse/commit/3ea955b270331cd1591c90723f7a047220bcb2e2))
- **package:** add commitlint to husky ([23ed444](https://github.com/JonDotsoy/envuse/commit/23ed444ad7c7a33bed3d1270b4396da7ce287897))
- **package:** upgrade scripts on packages ([c914ff3](https://github.com/JonDotsoy/envuse/commit/c914ff31490875a5ca2a791b6811f9265e33149a))

### [2.4.1](https://github.com/JonDotsoy/envuse/compare/envuse@v2.4.0...envuse@v2.4.1) (2021-07-28)

### Features

- **envuse:** upgrade script release ([5487f65](https://github.com/JonDotsoy/envuse/commit/5487f65a2dc74d5e9a7235915fd6646a5480bf5c))

## [2.4.0](https://github.com/JonDotsoy/envuse/compare/envuse@v2.3.0...envuse@v2.4.0) (2021-07-28)

### Features

- **envuse:** add file envuse-definition.ts ([2ec78bb](https://github.com/JonDotsoy/envuse/commit/2ec78bbb6e4a1fed8e213f954092ba52e275db95))

## [2.3.0](https://github.com/JonDotsoy/envuse/compare/envuse@v2.1.1...envuse@v2.3.0) (2021-07-27)

### Features

- add envuse lock file ([3091859](https://github.com/JonDotsoy/envuse/commit/30918590d717b771fbc53012c5de8997b2dde2d4))
- **cli:** configure legacy deps ([8b1fbdd](https://github.com/JonDotsoy/envuse/commit/8b1fbdd1aa6ba9fa2700542336c3be0453a0943a))
- **envuse:** added support to standard-version ([e901d18](https://github.com/JonDotsoy/envuse/commit/e901d185f6e05757ffcbce7f54fac2e03c836b34))
- **envuse:** upgrade inspect result ([d0bd146](https://github.com/JonDotsoy/envuse/commit/d0bd14680306d50c3083d780a617f762e9553db2))
- preare context envuse lock ([c4bfca2](https://github.com/JonDotsoy/envuse/commit/c4bfca2b0138f58127cfdc2153cba33825fd46a3))
- **vsce-envuse:** publish envuse.envuse-language@0.1.17 ([c7af40d](https://github.com/JonDotsoy/envuse/commit/c7af40d42f6bf26ff4d39bac800f7f80fa63ed0e))
