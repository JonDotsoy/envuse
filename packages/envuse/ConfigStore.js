const fs = require('fs');
const dotenv = require('dotenv');
const path = require('path');

class ConfigStore {
    constructor(init) {
        this.cwd = init.cwd;
    }

    getConfigs() {
        const cwd = this.cwd;
        return fs.readdirSync(`${cwd}`)
            .map(fileName => {
                const result = /^\.(?<config>.*)\.env$/i.exec(fileName);
                if (result) {
                    result.fullpath = path.resolve(`${cwd}/${result.input}`);
                    return result;
                }
            }).filter(Boolean);
    }

    find(configName) {
        return this.getConfigs().find(e => e.groups.config === configName);
    }

    getConfig(configName) {
        const config = this.find(configName);

        return dotenv.parse(fs.readFileSync(config.fullpath));
    }

    setConfig(configName, newConfig) {
        const buffers = Object.entries(newConfig).map(([key, value]) => {
            const newKey = key.replace(/\"|\=/ig, "");
            const newValue = value.replace(/\n/gi, '\\n');
            // const valueFormat = /\"|\n/ig.test(newValue) ? JSON.stringify(newValue): newValue;
            const valueFormat = newValue;
            
            return Buffer.from(`${newKey}=${valueFormat}\n`);
        });

        const buffer = Buffer.concat(buffers);

        const config = this.find(configName);
        fs.writeFileSync(config.fullpath, buffer);
    }
}

module.exports = ConfigStore;
