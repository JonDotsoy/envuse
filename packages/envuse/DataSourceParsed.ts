import { DataSource } from "./data-source/data-source";

export type DataSourceParsed = ReturnType<typeof DataSource.parse>;
