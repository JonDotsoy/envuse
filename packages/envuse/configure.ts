import { AsyncLocalStorage } from "async_hooks";

class Configuration {
  constructor(config?: Partial<Configuration>) {
    Object.assign(this, config);
  }
}

const globalConfiguration = new Configuration();

const configureStore = new AsyncLocalStorage<Configuration>();

export const getConfiguration = () =>
  configureStore.getStore() ?? globalConfiguration;

export const configure = (cb: (configurationLoad: Configuration) => void) =>
  cb(getConfiguration());

export const withConfiguration = <R>(
  config: Partial<Configuration>,
  cb: () => R
) => configureStore.run(new Configuration(config), cb);
