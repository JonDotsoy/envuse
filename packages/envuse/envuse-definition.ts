
/* Default interface for EnvuseDefinition */
export interface EnvuseDefinition {
  [k: string]: any;
}
