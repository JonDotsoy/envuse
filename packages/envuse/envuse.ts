import fs from "fs";
import { DataSource, Option, Values } from "./data-source/data-source";
import { BlockType } from "./data-source/statements/comps/Block";
import { SymbolEqual } from "./data-source/statements/comps/SymbolEqual";
import { Variable } from "./data-source/statements/comps/Variable";
import {
  createTypeScriptDefinition,
  EnvuseDefinition,
} from "./global-envuse-loaded";

const variableLoader: { [k: string]: any } = {};

// ensure filepath is a file or return null
const file = (filepath?: string) => {
  if (filepath && fs.existsSync(filepath) && fs.statSync(filepath).isFile()) {
    return filepath;
  }
  return null;
};

export const defaultFilepath =
  file(process.env.ENVUSE_FILE_PATH) ?? file(`${process.cwd()}/.envuse`);

export const defaultLockFilepath =
  file(process.env.ENVUSE_LOCK_FILE_PATH) ?? `${process.cwd()}/.envuse-lock`;

/**
 * Load file .envuse or file defined on process.env.ENVUSE_FILE_PATH and put
 * values in to process.env
 *
 * Make a lock file with the sample file and definition of .envuse file.
 *
 * @hidden
 */
export const register = (opts?: {
  filepath?: string;
  lockFilepath?: string;
}) => {
  const filepath = opts?.filepath ?? defaultFilepath;
  const lockFilepath = opts?.lockFilepath ?? defaultLockFilepath;

  if (filepath) {
    const r = DataSource.parseFile(filepath, process.env);
    const { definitions, parsed, ast } = r;

    global.envuseDataSourceParsed = r;

    for (const [key, { value }] of Object.entries(definitions)) {
      variableLoader[key] = value;
    }

    createTypeScriptDefinition(r);

    // Check if not there is a lock file
    if (!fs.existsSync(lockFilepath)) {
      const bodyLock: number[] = [...ast.body];
      ast.elementList.map((el) => {
        if (el instanceof Variable) {
          const pos =
            el.elementList.find((el) => el instanceof SymbolEqual)?.pos ??
            el.valueVariable.pos;
          bodyLock.fill(0, pos, el.valueVariable.end);
        }
      });
      fs.writeFileSync(
        lockFilepath,
        Buffer.from(bodyLock.filter((e) => e !== 0))
      );
    }

    for (const key in parsed) {
      if (parsed.hasOwnProperty(key)) {
        process.env[key] = parsed[key];
      }
    }
  }
};

export const parse = (option: Option, values?: Values) => {
  return DataSource.parse(option, values);
};

export const parseFile = (filepath: string, values: Values) => {
  return DataSource.parseFile(filepath, values);
};

export const createDataSource = (option: Option) => {
  return DataSource.createDataSource(option);
};

export const stringify = (comp: BlockType) => {
  return DataSource.stringify(comp);
};

export default variableLoader as EnvuseDefinition;
