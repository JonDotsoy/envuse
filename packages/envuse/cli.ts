#!/usr/bin/env node

import child_process from "child_process";
import envuse from ".";

const main = async () => {
  envuse.register();

  const [command, ...args] = process.argv.splice(2);

  if (!command) {
    throw new Error("No command specified");
  }

  child_process.spawnSync(command, args, {
    cwd: process.cwd(),
    env: process.env,
    stdio: "inherit",
    shell: true,
  });
};

main().catch((err) => {
  console.error(err);
  process.exit(1);
});
