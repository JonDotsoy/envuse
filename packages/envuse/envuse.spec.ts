import * as envuse from "./envuse";

describe("envuse", () => {
  it("shoud parse a buffer", () => {
    const buffer = Buffer.from("FOO=bar\nBAZ=qux\n", "utf8");
    const { parsed } = envuse.parse(buffer);
    expect(parsed).toEqual({
      FOO: "bar",
      BAZ: "qux",
    });
  });

  it("should create data source", () => {
    const buffer = Buffer.from("FOO=bar\nBAZ=qux\n", "utf8");
    const ast = envuse.createDataSource(buffer);
    expect(JSON.stringify(ast, null, 2)).toMatchInlineSnapshot(`
      "{
        \\"$type\\": \\"Block\\",
        \\"pos\\": 0,
        \\"end\\": 16,
        \\"children\\": [
          {
            \\"$type\\": \\"Variable\\",
            \\"pos\\": 0,
            \\"end\\": 7,
            \\"keyVariable\\": {
              \\"$type\\": \\"VariableKey\\",
              \\"pos\\": 0,
              \\"end\\": 3,
              \\"value\\": \\"FOO\\"
            },
            \\"valueVariable\\": {
              \\"$type\\": \\"VariableValue\\",
              \\"pos\\": 4,
              \\"end\\": 7,
              \\"value\\": \\"bar\\"
            }
          },
          {
            \\"$type\\": \\"SpaceNewLine\\",
            \\"pos\\": 7,
            \\"end\\": 8
          },
          {
            \\"$type\\": \\"Variable\\",
            \\"pos\\": 8,
            \\"end\\": 15,
            \\"keyVariable\\": {
              \\"$type\\": \\"VariableKey\\",
              \\"pos\\": 8,
              \\"end\\": 11,
              \\"value\\": \\"BAZ\\"
            },
            \\"valueVariable\\": {
              \\"$type\\": \\"VariableValue\\",
              \\"pos\\": 12,
              \\"end\\": 15,
              \\"value\\": \\"qux\\"
            }
          },
          {
            \\"$type\\": \\"SpaceNewLine\\",
            \\"pos\\": 15,
            \\"end\\": 16
          }
        ]
      }"
    `);
  });
});
