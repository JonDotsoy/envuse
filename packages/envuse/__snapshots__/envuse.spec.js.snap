// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`envuse should create data source 1`] = `
"{
  \\"$type\\": \\"Block\\",
  \\"pos\\": 0,
  \\"end\\": 16,
  \\"children\\": [
    {
      \\"$type\\": \\"Variable\\",
      \\"pos\\": 0,
      \\"end\\": 8,
      \\"keyVariable\\": {
        \\"$type\\": \\"VariableKey\\",
        \\"pos\\": 0,
        \\"end\\": 3,
        \\"value\\": \\"FOO\\"
      },
      \\"valueVariable\\": {
        \\"$type\\": \\"VariableValue\\",
        \\"pos\\": 4,
        \\"end\\": 7,
        \\"value\\": \\"bar\\"
      }
    },
    {
      \\"$type\\": \\"Variable\\",
      \\"pos\\": 8,
      \\"end\\": 16,
      \\"keyVariable\\": {
        \\"$type\\": \\"VariableKey\\",
        \\"pos\\": 8,
        \\"end\\": 11,
        \\"value\\": \\"BAZ\\"
      },
      \\"valueVariable\\": {
        \\"$type\\": \\"VariableValue\\",
        \\"pos\\": 12,
        \\"end\\": 15,
        \\"value\\": \\"qux\\"
      }
    }
  ]
}"
`;
